### PR TITLE
Fixed two bugs I added, an addExtras enhancement, updated example

### PR DESCRIPTION
**Pull request by nfriedly**
The primary reason for this is to fix the two bugs I introduced with my previous pull request:
- AwsSum.prototype.request once again properly handles errors on sending the request out
- It also once again accepts Buffers for the body. (This is useful for binary data.)

I didn't like how the only extra on `S3.PubObject` was the MD5 one which in turn called the content-length one. So I changed it to be an array containing both functions. Then I realized that it didn't matter because content-length was a required header so it would always go with the user-supplied one or else throw an error. But `S3.PutBucket` also needed both an MD5 and a Content-Length so I kept the changes.

I also changed `extrasContentLength` to use `Buffer.byteLength()` because both `string.length` and `buffer.length` can lead to incorrect content-length headers. It still won't work for non-string bodies, but it looks like it won't ever be called with a non-string body currently.

Lastly, I updated the s3 put-object example to use `Buffer.byteLength()` rather than have the content-length hard-coded and also commented to check the s3-streaming example for an example with a file.

So, in short, two bug fixes and a lot of junk :P

### CODE REVIEW

**Comment by chilts** [comment]
Hi Nathan,

I have cherry picked each of these commits and just fixed up the formatting on each of them (I don't use tabs in any code). :)

On the 2nd commit, I left the extrasContentLength() but I took it out of the addExtras for S3:PutBucketLifecycle since it's not actually required, even though the example has it in there. :)
- http://docs.amazonwebservices.com/AmazonS3/latest/API/RESTBucketPUTlifecycle.html

Again, many, many thanks for your contributions. I hope all of your Node.js and AWS stuff is going well.

Cheers,
Andy

**Comment by nfriedly** [comment]
Sounds good to me. I opened a feature request on node core for adding content-length to ReadableStreams because I think it would be incredibly useful for a library like this. But I'm not sure if it will actually make it in or not. https://github.com/joyent/node/issues/3144
